File: BEmu/BloombergTypes/Name.cs
//------------------------------------------------------------------------------
//  This code comes from Bloomberg.Blpapi.dll
//------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using System.Collections;
using System.Threading;

namespace Bloomberglp.Blpapi
{
    public class Name
    {
        private static Hashtable s_globalNameTable = new Hashtable();
		private static readonly object s_globalNameTableLock = new object();
		private static Random d_hashGenerator = new Random();
		private string d_name;
		private Name d_impl;
		private int d_hash;
		public static Name NullName = new Name(0);

		public Name()
		{
			this.d_name = Name.NullName.d_name;
			this.d_impl = Name.NullName.d_impl;
			this.d_hash = Name.NullName.d_hash;
		}
		private Name(int zero)
		{
			this.d_name = null;
			this.d_impl = this;
			this.d_hash = Name.d_hashGenerator.Next();
		}
		public Name(string nameString)
        {
            Name name = Name.GetName(nameString);
            this.d_impl = name;
            this.d_name = this.d_impl.d_name;
            this.d_hash = this.d_impl.d_hash;
		}
        private Name(string nameString, bool unUsed)
		{
			this.d_name = nameString;
			this.d_impl = this;
            Name.s_globalNameTable[nameString] = this;
            this.d_hash = Name.d_hashGenerator.Next();
		}
        public static Name GetName(string nameString)
		{
			if (nameString == null)
			{
                return Name.NullName;
			}
			object obj;
            Monitor.Enter(obj = Name.s_globalNameTableLock);
			Name result;
			try
			{
                Name name = (Name)Name.s_globalNameTable[nameString];
				if (name == null)
				{
                    name = new Name(nameString, true);
				}
				result = name;
			}
			finally
			{
				Monitor.Exit(obj);
			}
			return result;
		}
        public static Name FindName(string nameString)
		{
			if (nameString == null)
			{
                return Name.NullName;
			}
			object obj;
            Monitor.Enter(obj = Name.s_globalNameTableLock);
            Name result;
			try
			{
                result = (Name)Name.s_globalNameTable[nameString];
			}
			finally
			{
				Monitor.Exit(obj);
			}
			return result;
		}
		public static bool HasName(string nameString)
		{
			if (nameString == null)
			{
				return true;
			}
			object obj;
            Monitor.Enter(obj = Name.s_globalNameTableLock);
			bool result;
			try
			{
                result = Name.s_globalNameTable.ContainsKey(nameString);
			}
			finally
			{
				Monitor.Exit(obj);
			}
			return result;
		}
		public override bool Equals(object other)
		{
			if (other is string)
			{
				return this.d_name.Equals(other);
			}
            return other is Name && ((Name)other).d_impl == this.d_impl;
		}
		public override int GetHashCode()
		{
			return this.d_hash;
		}
		public override string ToString()
		{
			return this.d_name;
		}
		public static void FreeGlobalNameTable()
		{
			object obj;
            Monitor.Enter(obj = Name.s_globalNameTableLock);
			try
			{
                Name.s_globalNameTable.Clear();
			}
			finally
			{
				Monitor.Exit(obj);
			}
		}
    }
}

File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/Name.java
//------------------------------------------------------------------------------
//	This code comes from blpapi-3.6.1-0.jar
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import java.util.HashMap;

public final class Name
{
	private static HashMap<String, Name> s_globalNameTable = new HashMap<String, Name>();
	private static final Object s_globalNameTableLock = new Object();
	private final String d_name;
	private final Name d_impl;

	public static final Name NullName = new Name(0);
	
	public Name()
	{
		this.d_name = NullName.d_name;
		this.d_impl = NullName.d_impl;
	}
	
	private Name(int zero)
	{
		assert (zero == 0);
		this.d_name = null;
		this.d_impl = this;
	}
	
	public Name(String nameString)
	{
		Name localName = getName(nameString);
		this.d_impl = localName;
		this.d_name = this.d_impl.d_name;
	}
	
	private Name(String nameString, boolean unUsed)
	{
		this.d_name = nameString;
		this.d_impl = this;
		s_globalNameTable.put(nameString, this);
	}
	
	public static Name getName(String nameString)
	{
		if (nameString == null) {
			return NullName;
		}
		synchronized (s_globalNameTableLock)
		{
			Name localName = (Name)s_globalNameTable.get(nameString);
			if (localName == null) {
				localName = new Name(nameString, true);
			}
			return localName;
		}
	}
	
	public static Name findName(String nameString)
	{
		if (nameString == null) {
			return NullName;
		}
		synchronized (s_globalNameTableLock)
		{
			return (Name)s_globalNameTable.get(nameString);
		}
	}
	
	public static boolean hasName(String nameString)
	{
		if (nameString == null) {
			return true;
		}
		synchronized (s_globalNameTableLock)
		{
			return s_globalNameTable.containsKey(nameString);
		}
	}
	
	public boolean equals(Object other)
	{
		if ((other instanceof String)) {
			return this.d_name.equals(other);
		}
		if ((other instanceof Name)) {
			return ((Name)other).d_impl == this.d_impl;
		}
		return false;
	}
	
	public int hashCode()
	{
		if (this == this.d_impl) {
			return super.hashCode();
		}
		return this.d_impl.hashCode();
	}
	
	public String toString()
	{
		return this.d_name;
	}
	
}
